### ISSUE TITLE
Fused foreach lerp

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#71683]
### 🚀 The feature, motivation and pitch

I'd like to simplify the exponential moving average optimization update for teacher-student model:
```python
for param_teacher, param_student in zip(self.teacher_encoder_proj.parameters(), self.student_encoder_proj.parameters()):
           param_teacher.data.lerp_(1 - param_momentum, param_student.data)
```

I was looking for a foreach version of lerp used in https://github.com/pytorch/pytorch/blob/acb340de75367985164a67234be7ee7ddf4b546e/torch/optim/_multi_tensor/_functional.py and found only :

```python
torch._foreach_mul_(exp_avg_sq, beta2)
torch._foreach_addcmul_(exp_avg_sq, grads, grads, 1 - beta2)
```

Not sure if this is indeed lerp, but if it is, it would be beneficial to have a `_foreach_lerp_` that has a simpler API and saves the number of kernel calls.

It would also be nice to have the available foreach method list documented somewhere even with a notice that they are experimental and are subject to change: https://pytorch.org/docs/stable/search.html?q=foreach&check_keywords=yes&area=default#

**Comment by vadimkantorov** [comment]
It may also be a good idea to introduce an EMA "optimizer" (at least the functional version), to make sure parameters match between teacher and student model, it may be a good idea to force users to provide  `param_names` list inside the param group

**Comment by vadimkantorov** [comment]
Related vision issues:

* https://github.com/pytorch/vision/issues/5284
* https://github.com/pytorch/vision/blob/c8e3b2a5925e7b7ed21662e86a7e9553170a5633/references/classification/utils.py#L175-L184

**Comment by vadimkantorov** [comment]
My current impl:
```python
def update_param_groups(param_groups, key, value):
    value = torch.as_tensor(value)
    values = value.expand(len(param_groups)) if len(value) != len(param_groups) else value
    for g, v in zip(param_groups, values.tolist()):
        g[k] = v

class MeanTeacherMomentumScheduler:
    def __init__(self, optimizer, warmup_steps = 20_000):
        self.optimizer = optimizer
        self.warmup_steps = warmup_steps

    def step(self, iteration):
        momentum = 0.0 if iteration <= self.warmup_steps else self.optimizer.defaults['momentum']
        update_param_groups(self.optimizer.param_groups, 'otpimizer', momentum)

class MeanTeacher(torch.optim.Optimizer):
    def __init__(self, param_groups, optimizer, momentum = 0.999):
        super().__init__(params = param_groups, defaults = dict(momentum = momentum))
        self.optimizer = optimizer
        self._param_groups = self.param_groups

        self.param_groups = optimizer.param_groups
        self.zero_grad = optimizer.zero_grad

    def step(self):
        self.optimizer.step()
        self.ema_(teacher_param_groups = self._param_groups, student_param_groups = self.param_groups)

    @staticmethod
    @torch.no_grad()
    def ema_(teacher_param_groups, student_param_groups, momentum = None):
        student_params = sum([g['params'] for g in student_param_groups], [])
        teacher_params = sum([g['params'] for g in teacher_param_groups], [])
        teacher_momentums = sum([[g['momentum']] * len(g['params']) for g in teacher_param_groups], []) if momentum is None else [momentum] * len(teacher_params)
        for pt, ps, momentum in zip(teacher_params, student_params, teacher_momentums):
            if momentum == 0:
                pt.copy_(ps)
            elif momentum < 1:
                pt.lerp_(ps, 1 - momentum)
```

related lerp issue: https://github.com/pytorch/pytorch/issues/71701

**Comment by vadimkantorov** [comment]
Related: https://github.com/pytorch/pytorch/commit/7a6c4d0c50dd0670d87bc39d53292cf8cb90ca04#diff-2673a2f4bab81258173df4e5cdf7f678b691dceb8045f1d9761552576d0e1e8aR55

**Comment by vadimkantorov** [comment]
Regarding https://github.com/pytorch/pytorch/pull/87562#issuecomment-1375597894: I propose a function `mix3 = lambda a = 1, b = 1, c = 1, A = 1, B = 1, C = 1: a*b*c + A*B*C` where all a,b,c,A,B,C can be scalars or tensors. An inplace mode of this function could then be used to idiomatically compute moments second moments in optimizers, as proposed in https://github.com/pytorch/pytorch/issues/79352#issuecomment-1152946076

**Comment by vadimkantorov** [comment]
foreach_lerp support is soon merged, so I repurposed this issue a bit to be a feature request for an EMA optimizer: class-form and function-form using the new foreach_lerp

**Comment by vadimkantorov** [comment]
An example of how explicit BatchNorm running stats could look like:
https://github.com/pytorch/pytorch/pull/94820#issuecomment-1431373114
